### PR TITLE
 prov/rxm: Fix SAR segments receive handling

### PR DESCRIPTION
**Pull request by dmitrygladkov**
Check all items in unexpected queue that can be matched with new posted
recv entry. It allows to handle situation when a rank is waiting for the
message with id=1 (transfered by SAR protocol). But due to re-tansmisisions
it is highly possible that this rank (receiver) can receive new SAR segments
from another ranks.
Also adds checks for connection to avoid situation when RxM reads from another
transfer to the current tranfer that is underr processing.

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
I've checked this with mpi_stress tool with TCP and Verbs providers (works fine even if set FI_OFI_RXM_MSG_TX|RX_SIZE=8). @a-ilango, if you have enough time, it would be great to double-check on your end and be sure that everything is fine.

File: prov/rxm/src/rxm.h [hunk 1]
```diff
@@ -701,28 +707,46 @@ rxm_process_recv_entry(struct rxm_recv_queue *recv_queue,
...
 		} else {
+			struct dlist_entry *entry;
 			enum rxm_sar_seg_type last =
 				(rxm_sar_get_seg_type(&rx_buf->pkt.ctrl_hdr)
 								== RXM_SAR_SEG_LAST);
 			ssize_t ret = rxm_cq_handle_rx_buf(rx_buf);
+			struct rxm_recv_match_attr match_attr = {
+				.addr	= recv_entry->addr,
+				.tag	= recv_entry->tag,
+				.ignore	= recv_entry->ignore,
```
Comment (shefty) [inline prov/rxm/src/rxm.h]: This initializes match_attr right before we might return from the call.  This would be better moved down past the if check.

Comment (dmitrygladkov) [reply prov/rxm/src/rxm.h]: I agree. Done

**Comment by dmitrygladkov** [comment]
> We may want to refactor the code later if possible. Checking the unexpected queue twice (in the code) looks odd.

Yes I agree, that we need to refactor this, because the logic of SAR is too complicated.

**Pull request merged**